### PR TITLE
fix (core): EffectInstaKill Prevents spell animation fix

### PR DESCRIPTION
**Pull request by acidmanifesto**
This fixes the issue of spell animations of chain effects of a spell dealing with EffectInstaKill. This was identified as a issue in support of @malcrom 's current scripting efforts. We fix the issue by putting a half second delay and making the delay of a function possible and independent.

Co-Authored-By: Malcrom <malcromdev@gmail.com>

## Changes Proposed:
-  Add independent Delay Feature in SpellEffect.cpp

## Issues Addressed:

- Closes 

## SOURCE:

Sniff Team

## Tests Performed:

- Builds
- Performs

## How to Test the Changes:

1. Currently eith the pr do a .cast back 33974 on a creature and you will see the power burn happen follow by a bomb explosion (shockwave effect) then instakill.
2. without this pr you will not see the bomb explosion shockwave effect

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by malcrom — APPROVED** [review]

File: src/server/game/Spells/SpellEffects.cpp [hunk 1]
```diff
@@ -61,13 +61,18 @@
...
+#include <iostream>
+#include <ctime>
 
  // TODO: this import is not necessary for compilation and marked as unused by the IDE
 //  however, for some reasons removing it would cause a damn linking issue
 //  there is probably some underlying problem with imports which should properly addressed
 //  see: https://github.com/azerothcore/azerothcore-wotlk/issues/9766
 #include "GridNotifiersImpl.h"
 
+using namespace std;
```
Comment (Nefertumm) [inline src/server/game/Spells/SpellEffects.cpp:74]: Remove the namespace please :D

Comment (acidmanifesto) [reply src/server/game/Spells/SpellEffects.cpp]: hopefully it compiles. I put the std in STuD and all I need is U.

File: src/server/game/Spells/SpellEffects.cpp [hunk 2]
```diff
@@ -61,6 +61,9 @@
 #include "Vehicle.h"
 #include "World.h"
 #include "WorldPacket.h"
+#include <chrono>
+#include <iostream>
+#include <ctime>
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellEffects.cpp:66]:
```suggestion
#include <chrono>
#include <ctime>
#include <iostream>
```

File: src/server/game/Spells/SpellEffects.cpp (hunk 1 above)
Comment (Nefertumm) [reply src/server/game/Spells/SpellEffects.cpp]: Probably it won't, you'll need to add std:: before the things :P

File: src/server/game/Spells/SpellEffects.cpp (hunk 2 above)
Comment (acidmanifesto) [reply src/server/game/Spells/SpellEffects.cpp:66]: i thought ct came before ch? lol

File: src/server/game/Spells/SpellEffects.cpp [hunk 3]
```diff
@@ -273,6 +276,15 @@ void Spell::EffectResurrectNew(SpellEffIndex effIndex)
     SendResurrectRequest(target);
 }
 
+void DelayMe(float seconds)
+{
+    clock_t startClock = clock();
+    float secondsAhead = seconds * CLOCKS_PER_SEC;
+    // do nothing until the elapsed time has passed.
+    while (clock() < startClock + secondsAhead);
```
Comment (Nefertumm) [inline src/server/game/Spells/SpellEffects.cpp:284]:
```suggestion
    std::clock_t startClock = std::clock();
    float secondsAhead = seconds * CLOCKS_PER_SEC;
    // do nothing until the elapsed time has passed.
    while (std::clock() < startClock + secondsAhead);
```
If I'm not missing something :P

Comment (acidmanifesto) [reply src/server/game/Spells/SpellEffects.cpp:284]: D: gasp. addded in

File: src/server/game/Spells/SpellEffects.cpp (hunk 1 above)
Comment (acidmanifesto) [reply src/server/game/Spells/SpellEffects.cpp]: it compiles :D

**Review by Nefertumm — APPROVED** [review]

File: src/server/game/Spells/SpellEffects.cpp [hunk 4]
```diff
@@ -273,6 +276,15 @@ void Spell::EffectResurrectNew(SpellEffIndex effIndex)
     SendResurrectRequest(target);
 }
 
+void DelayMe(float seconds)
+{
+    std::clock_t startClock = std::clock();
+    float secondsAhead = seconds * CLOCKS_PER_SEC;
+    // do nothing until the elapsed time has passed.
+    while (std::clock() < startClock + secondsAhead);
+    return;
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/SpellEffects.cpp:285]: no no no. I'll try to do it differently

File: src/server/game/Spells/SpellEffects.cpp [hunk 5]
```diff
@@ -273,6 +276,15 @@ void Spell::EffectResurrectNew(SpellEffIndex effIndex)
...
+}
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/SpellEffects.cpp:286]:
```suggestion
void DelayMe(Milliseconds delayMs)
{
    SystemTimePoint delay = std::chrono::system_clock::now() + delayMs;

    // do nothing until the elapsed time has passed.
    while (std::chrono::system_clock::now() < delay);

    return;
}
```

File: src/server/game/Spells/SpellEffects.cpp [hunk 6]
```diff
@@ -61,6 +61,9 @@
 #include "Vehicle.h"
 #include "World.h"
 #include "WorldPacket.h"
+#include <chrono>
+#include <ctime>
+#include <iostream>
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/SpellEffects.cpp:66]:
```suggestion
```

File: src/server/game/Spells/SpellEffects.cpp [hunk 7]
```diff
@@ -294,6 +306,8 @@ void Spell::EffectInstaKill(SpellEffIndex /*effIndex*/)
     data << uint32(m_spellInfo->Id);
     m_caster->SendMessageToSet(&data, true);
 
+    DelayMe(0.5f);
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/SpellEffects.cpp:309]:
```suggestion
    DelayMe(500ms);
```

File: src/server/game/Spells/SpellEffects.cpp (hunk 5 above)
Comment (Kitzunu) [reply src/server/game/Spells/SpellEffects.cpp:286]: Nice catch, for a second I totally forgot about chrono

**Comment by Kitzunu** [comment]
DelayMe...

Can we give it a better name? Can it be used in more places. I feel like just DelayMe isnt describing it too well without knowing if this is the only place it will only be used

**Comment by Nyeriah** [comment]
So this is basically https://www.cplusplus.com/reference/thread/this_thread/sleep_for/

**Comment by Kitzunu** [comment]
Well yes, but as it is introduced in SpellEffects, am I right to believe that this function will never serve a purpose anywhere else? Or should we move the function to a more generic place?

**Comment by Nyeriah** [comment]
> Well yes, but as it is introduced in SpellEffects, am I right to believe that this function will never serve a purpose anywhere else? Or should we move the function to a more generic place?

Yes, that's what I meant, it can be used elsewhere if we want to, I just would recomend against replacing the usual schedulers (eventmap, taskcheduler) with this, if an action is supposed to be simply delayed. Just so people don't start doing weird things.

**Review by Kitzunu — CHANGES_REQUESTED** [review]
Please address what was discussed above.

**Comment by acidmanifesto** [comment]
> Well yes, but as it is introduced in SpellEffects, am I right to believe that this function will never serve a purpose anywhere else? Or should we move the function to a more generic place?

Oh more discussions last night this function will most likely be used in another spot after a debugging session of another issue.

**Comment by Kitzunu** [comment]
better names can be discussed, I think Nye suggested it due to https://www.cplusplus.com/reference/thread/this_thread/sleep_for/

**Comment by acidmanifesto** [comment]
Maybe DeLaY_FunCtiON or something that will drive the ocd wild.

**Comment by Nyeriah** [comment]
I suggested sleep_for because this function does exactly the same thing so it would probably be more intuitive for developers. But other names could be used.

**Comment by Kitzunu** [comment]
Maybe a good place for it is in Util, or it's own file

```cpp
void Acore::sleep_for(Milliseconds delayMs)
{
    SystemTimePoint delay = std::chrono::system_clock::now() + delayMs;

    // do nothing until the elapsed time has passed.
    while (std::chrono::system_clock::now() < delay);
}
```

```cpp
namespace Acore
{
    AC_COMMON_API void sleep_for(Milliseconds delayMs);
}
```

**Comment by Kitzunu** [comment]
Care to elaborate?

**Comment by acidmanifesto** [comment]
> I suggested sleep_for because this function does exactly the same thing so it would probably be more intuitive for developers. But other names could be used. 

I was obviously kidding about the ocd thing but sleep_for is a pretty good idea. We can go for that since thats some form of coding standard of a naming convention as you pointed out.

**Comment by acidmanifesto** [comment]
> Care to elaborate?

Yes

File: src/server/game/Spells/SpellEffects.cpp [hunk 8]
```diff
@@ -273,6 +273,14 @@ void Spell::EffectResurrectNew(SpellEffIndex effIndex)
...
 }
 
+void DelayMe(Milliseconds delayMs)
+{
+    SystemTimePoint delay = std::chrono::system_clock::now() + delayMs;
+
+    // do nothing until the elapsed time has passed.
+    while (std::chrono::system_clock::now() < delay);
+}
+
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/SpellEffects.cpp:283]:
```suggestion
```

File: src/server/game/Spells/SpellEffects.cpp [hunk 9]
```diff
@@ -294,6 +302,8 @@ void Spell::EffectInstaKill(SpellEffIndex /*effIndex*/)
     data << uint32(m_spellInfo->Id);
     m_caster->SendMessageToSet(&data, true);
 
+    DelayMe(500ms);
```
[the remark this answers is not in the document]
Comment (Kitzunu) [reply src/server/game/Spells/SpellEffects.cpp:305]: I guess that works too

**Comment by Winfidonarleyan** [comment]
> Care to elaborate?

Need use https://en.cppreference.com/w/cpp/thread/sleep_for

**Comment by Kitzunu** [comment]
> > Care to elaborate?
> 
> Need use https://en.cppreference.com/w/cpp/thread/sleep_for

Nice, then we don't have to implement anything.

**Comment by Nyeriah** [comment]
I'm not really a threading person so not sure if that will lead us to the expected results, but one can always try xD

**Comment by Winfidonarleyan** [comment]
> I'm not really a threading person so not sure if that will lead us to the expected results, but one can always try xD

It will be the same as it is now.

**Review by Winfidonarleyan — APPROVED** [review]

File: src/server/game/Spells/SpellEffects.cpp [hunk 10]
```diff
@@ -294,6 +294,8 @@ void Spell::EffectInstaKill(SpellEffIndex /*effIndex*/)
     data << uint32(m_spellInfo->Id);
     m_caster->SendMessageToSet(&data, true);
 
+    std::this_thread::sleep_for(500ms);
```
[the remark this answers is not in the document]
Comment (acidmanifesto) [reply src/server/game/Spells/SpellEffects.cpp:297]: i will do a local compile first and see where that takes me. so i will check back if that include is needed or not.

**Review by Kitzunu — APPROVED** [review]

File: src/server/game/Spells/SpellEffects.cpp (hunk 10 above)
Comment (Kitzunu) [inline src/server/game/Spells/SpellEffects.cpp:297]: actually, can you add a comment to this why we need it.

**Comment by Alastro-Kraken** [comment]
Hello! Just a note - this is a broader issue than just instakill spells. Whenever a player uses a spell to kill a mob, the mob's death is processed instantly by the server and spell animations are aborted. It's easy to test with feral druids' Mangle because it has a clear spell effect.
I would say that the client expects some delay as per the packet batching system that was active under WotLK, although I have no research to back that up.
If packet batching is a no-go, then maybe delay unit deaths by ~400ms instead?

**Comment by acidmanifesto** [comment]
> Hello! Just a note - this is a broader issue than just instakill spells. Whenever a player uses a spell to kill a mob, the mob's death is processed instantly by the server and spell animations are aborted. It's easy to test with feral druids' Mangle because it has a clear spell effect. I would say that the client expects some delay as per the packet batching system that was active under WotLK, although I have no research to back that up. If packet batching is a no-go, then maybe delay unit deaths by ~400ms instead?

Thank you for the detail and info. I was able to identify where we needed it ot be applied and got it working via mangle showing all effects on creature death now.  https://github.com/azerothcore/azerothcore-wotlk/pull/10394

File: src/server/game/Spells/SpellEffects.cpp [hunk 11]
```diff
@@ -293,6 +293,9 @@ void Spell::EffectInstaKill(SpellEffIndex /*effIndex*/)
     data << unitTarget->GetGUID();
     data << uint32(m_spellInfo->Id);
     m_caster->SendMessageToSet(&data, true);
+    // MDic - acidmanifesto: This is needed due to the Unit::DealDamage firing off immediately after the worldpacket which nerfs out additional spell animations
+    // The delay of half a second ensures the spell animations which are typically 0.4 seconds long have enough time to fire off
+    std::this_thread::sleep_for(500ms);
```
Comment (jameyboor) [inline src/server/game/Spells/SpellEffects.cpp:298]: How did this ever get accepted?

**Comment by Kitzunu** [comment]
Hey!

Thanks for your feedback, but this was already reverted. I'll lock this PR now because it seems to be misleading too many people.

For reference, it was reverted here: 8f249a3

If you are encountering issues, please update your core.

A new solution to this issue will be presented soon, make sure to check it out once it's ready.

Thanks once again!

**Pull request merged**